Refactor Lightcurve class and add more features

OrkoHunter opened the pull request:
We want to expand the Lightcurve class to add new features and properties to it. I have started with drawing feature.

Usage :
```python
>>> import matplotlib.pyplot as plt
>>> import stingray
>>> # Create time and counts array
>>> lc = stingray.Lightcurve(time, counts)
>>> lc.draw()
>>> plt.show()
```

nithinsingh61 commented:
you have lightcurve object ?

OrkoHunter commented:
I don't have it. I created random time and counts array to test.
On 10 Mar 2016 10:36 am, "Nithin singh" <notifications@github.com> wrote:

> you have lightcurve object ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/StingraySoftware/stingray/pull/74#issuecomment-194667971>
> .
>

nithinsingh61 commented:
Great we should be doing this ! I will even add those  after finishing up with unit tests

OrkoHunter commented:
Awesome !
On 10 Mar 2016 10:42 am, "Nithin singh" <notifications@github.com> wrote:

> Great we should be doing this ! I will even add those after finishing up
> with unit tests
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/StingraySoftware/stingray/pull/74#issuecomment-194671421>
> .
>

OrkoHunter commented:
I was wondering if it makes sense to add two Lighcurve objects. We can add them if they have the same bin type and size. So, we could be adding in one of the two ways :

1. Concatenation : It will be similar to extending the dataset of one over the other.

2. Summing up : Adding individual values per bin. I won't go for this though because it doesn't seem to make much sense. Or I might be wrong.

usmanwardag commented:
How do we know bin type? Is there any attribute in the `lightcurve` class that defines it?

OrkoHunter commented:
@usmanwardag Sorry for being vague. By type, I meant the boolean `input_counts`.

matteobachetti commented:
It would be very useful, and Maltpynt does that in fact. Check
join_lightcurves and scrunch_lightcurves to see a way to do that (and
evolve it if you have different ideas :) )

On 17:02, Thu, Mar 10, 2016 Himanshu Mishra <notifications@github.com>
wrote:

> @usmanwardag <https://github.com/usmanwardag> Sorry for being vague. By
> type, I meant the boolean input_counts.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/StingraySoftware/stingray/pull/74#issuecomment-194924947>
> .
>

dhuppenkothen commented:
I like the drawing method! 
Two suggestions: 

* It would be great if the Figure could also show the measurement uncertainties (depends on #24)
* It would also be great to have maybe an option to save the Figure to a file?

OrkoHunter commented:
GitHub reactions ! [Thank you GitHub](https://github.com/dear-github/dear-github) :)

OrkoHunter commented:
Added function to save the figure.

OrkoHunter commented:
@dhuppenkothen Have a look now? I merged the `draw_and_savefig` in the `draw` itself.

OrkoHunter commented:
Hi ! I have added a method to join two light curves (Maybe like Maltpynt's scruntch_lightcurves). I have also added an example. Please see if that looks okay.

OrkoHunter commented:
I also think of calling the `join` method from `__add__`, so that
```python
>>> lc = lc1 + lc2
```
could do the job of `lc1.join(lc2)`.

matteobachetti commented:
Hi,
implementing a `join` method to concatenate the light curves is good!
But I'm absolutely *against* overriding `__add__`. This should be to sum light curves element by element, as default for numpy arrays, not the concatenation.
There are cases where you need to sum light curves element by element, and I'm strongly for maintaining this default behavior.

OrkoHunter commented:
Hmm.. That makes sense to me now. So we are not overriding `__add__` with `join`, but we still need to write the method for adding the curves element by element. I'm not completely sure how that can be done, but this is what I think

1. `lc1` and `lc2` must have same time arrays. (?)
2. Corresponding counts arrays will be simply added like default array addition.

Please suggest me how can I proceed. Thanks.

matteobachetti commented:
Yes, exactly as you say it. Check on time array, then normal array sum as in numpy arrays. Some of the `Lightcurve` attributes my need an update too (like `countrate`)

OrkoHunter commented:
I have a doubt, Why do we not check if `len(lc.time) == self.ncounts` ? If there are extra values in the counts array or the time array, what would they represent ? We can't say that they are missing data from between the start and end time, because they will always be at the ends.

OrkoHunter commented:
@dhuppenkothen @matteobachetti Please review this PR.

OrkoHunter commented on stingray/lightcurve.py:
```diff
@@ -175,3 +209,165 @@ def rebin_lightcurve(self, dt_new, method='sum'):
+        assert self.dt == other.dt, "The bin width of both the lightcurves" \
+                                    " must be same."
+        # TODO : If different, rebin both the lightcurves with
+        # dt_new = LCM(self.dt, other.dt) and raise a warning.
```
Should we go for this?

evandromr replied:
I don't see a real application of this behavior. I believe this case should raise an `Error` and not join, or raise a `Warning` and drop one of the duplicated time bin, in this case I would choose to keep the duplicate of lc1 (where the method is called) and drop the duplicate of lc2, though physically it also doesn't make sense to me so I would choose `Error`.

On the other hand this is the kind of operation that I would expect when summing light curves that do not completely overlap, but I would still expect a `Warning` from `SIMON`

OrkoHunter replied:
So, I think I should leave it like this and expect to get equal `dt` of both the light curves. Thank you for your inputs !

dhuppenkothen commented on stingray/lightcurve.py:
```diff
@@ -175,3 +209,163 @@ def rebin_lightcurve(self, dt_new, method='sum'):
+        """
+        assert self.dt == other.dt, "The bin width of both the lightcurves" \
+                                    " must be same."
+
```
I think analogously to #78, this should probably throw a warning rather than an error. There's no intrinsic reason why the bin widths of the two time series must be the same, except that it'll screw up some of the Fourier methods. However, we might include methods in the future where this doesn't matter, and it might still be useful, for example, for visualization on the GUI.

OrkoHunter replied:
I agree with you on this.

dhuppenkothen commented on stingray/lightcurve.py:
```diff
@@ -175,3 +209,163 @@ def rebin_lightcurve(self, dt_new, method='sum'):
+            utils.simon("use ``draw`` function to draw the image first and "
+                        "then use ``savefig`` to save the figure.")
+
+        plt.savefig(filename, **kwargs)
```
I would probably pull this method out of the class and stick it into `io.py` or `utils.py` or something, since it's general enough that we might want to use it elsewhere?

OrkoHunter replied:
Correct ! `draw` can vary from class to class but `savefig` would remain more or less same for all of the plots. I think we should move it to the `io` module.

OrkoHunter commented:
@dhuppenkothen I have addressed the last comments. If you can give some light on [this comment](https://github.com/StingraySoftware/stingray/pull/74#issuecomment-195711292) too, I think this would be good to go.

dhuppenkothen commented:
Ah! It took me a bit to understand what you mean, but I get it now. 
Yes, I agree, at the very beginning of `Lightcurve.__init__()` there should be a check if `time` and `counts` are the same length.

OrkoHunter commented:
Do we want to raise a warning or close with an error?

dhuppenkothen commented:
I think in this case, closing with an error seems like the correct behaviour. As you said, we really know what it means when time and counts aren't the same length, so I'd rather the user figured this out and we didn't try to guess.

OrkoHunter commented:
@dhuppenkothen Please see this now.

evandromr commented:
I still think that the `join` method is behaving how I would expect the `__add__` method to behave. It's basically padding the light curves with zeros to be of the same size and adding them element-wise. As it is, I believe those methods should be the same and raise different `Warning` messages for the different cases.

If the idea of the `join` methods is just to concatenate, performing a sum of counts element wise is not what I would expect. I believe the `join` method should check if the time arrays of the light curves do not overlap and, if they do, warn the user and drop one of the duplicate bins, but not sum the counts. 

The warning message could be something like:
> The light curves you are joining have overlapping time ranges, we are keeping the original values in those ranges. If you wish to sum, please use `lc_sum = lc1 + lc2`

("original value" can be ambiguous, there is probably a clearer way to formulate this)

**Sumarizing my opinion:**
  - `join` should not add, `__add__` should add.
  - light curves do not have to have the same time array to be summed, they can be padded by zeros.(Warning should be raised).
  - Concatenating light curves don't have to have the same time bin width. (Warning should be raised).

Can someone try to convince me otherwise? :wink:

matteobachetti commented:
For the use we generally do of light curves, using `join` should *never* get to overlap.
An overlap should raise a warning, if not an exception. If we choose to be soft about it, maybe we could average the overlapping bins instead of summing or dropping.
In any case, we should account for possible zeroed-out bins in a proper GTI array.

OrkoHunter commented:
Hey @evandromr @matteobachetti , Sorry for being late on this. I changed the `join` method which warns the user in case of overlapping time arrays and adds the average value to the resulting count array. Does this look good?

evandromr commented on stingray/lightcurve.py:
```diff
@@ -221,6 +221,8 @@ def join(self, other):
         objects. The individual count per bin in the resulting object will
         be the sum total of the count per bin in both lightcurve objects.
```
This sentence  is misleading. The result count per bin will be the individual count per bin, or the average in case of overlapping.

OrkoHunter replied:
Yes ! The sentence was wrong infact. Changed it.

OrkoHunter commented:
I don't know how coveralls calculate the coverage. I can't seem to find more tests to add.
I think this PR needs a final review. @matteobachetti @dhuppenkothen @evandromr

matteobachetti commented:
Hi @OrkoHunter, the `draw` method is what is pushing down the coverage stats for the lightcurve part, and `savefig` in `io.py`
We should definitely cover those, and have in `.travis.yml` at least a matrix point with no matplotlib *and* test coverage. That should fix the test coverage decrease

OrkoHunter commented:
I don't know how can we cover a `draw` and `savefig` functions. For `draw` we can at least test if the image has been drawn by `plt.fignum_exists` function but we can't be sure about what actually has been plot (but I guess that is covered by matplotlib).
I have added more tests to cover those methods.

matteobachetti commented:
Well, first of all you test that the function executes and does not crash; then, you can test that if matplotlib is not installed the correct error/warning is thrown. With the different combinations of libraries and Python version, this is already something not trivial.

OrkoHunter commented:
There's some problem with installing matplotlib on travis, whereas it is installed on Appveyor.

nithinsingh61 commented:
Had a similar problem  .Try not allowing  matplotlib to use x windows backend ( though the error persisted in mine due to PEP8 :(  )

OrkoHunter commented:
@nithinsingh61 Don't we first need matplotlib installed to change its backend ? (on travis)

nithinsingh61 commented:
I don't think installation is causing the problem . it is you trying to plot it in " travis environment " that is causing the problem

OrkoHunter commented:
@nithinsingh61 We can now add matplotlib tests using [this commit](https://github.com/StingraySoftware/stingray/pull/74/commits/cc5c250417440e48a629374f2396e7d863f22f6c). I found it in the travis docs itself.
Ping @matteobachetti @dhuppenkothen

nithinsingh61 commented:
Can you please consider renaming draw to plot as a new user to stingray who is accustomed to matplot may find it ambiguous  :) @OrkoHunter

OrkoHunter commented:
@nithinsingh61 I renamed the draw method to be consistent with matplotlib.

OrkoHunter commented:
I have added few more methods based on #90. I have not yet added tests yet but the doc tests should pass.

I want to discuss the philosophy of some of these methods here -

* __Sorting__ : Do we need to sort in place or we want to return a new object? I think disturbing the old object is not a very good idea and hence I've returned a new object with sorted counts array and time bins arranged accordingly. It raises a warning because the constant bin size is not maintained.

* __Indexing__ : What is the expected output one does `lc[1]` ? Maintaining consistency with array indexing I have currently returned the count value at _index 1_. But I feel there should be another method for getting the corresponding count value if we give it a time value. Something like `lc.count_at(10)` which gives count value at `t==10`. What are other opinions about this?

* __Slicing__ : Similar point as I raised for indexing as both comes follows method.

* __Truncation__ : I have added a method to truncate similar to slicing. But the doubt is that whether we should ask for the indices in the argument or actual time values. I have currently kept it strictly like normal python conventions.

matteobachetti commented:
@OrkoHunter these are very good suggestions. I'd vote to make `lc.sort()` a method that makes the sorting in place, while having `sorted(lc)` return a new object, for consistency with numpy arrays. For slicing and indexing, I agree on the `_at` version (or some other convention we agree on) that filters by time instead of index, while keeping indexing done by bin number.

OrkoHunter commented:
We have an issue here. There is no warning raised for [this test](https://travis-ci.org/StingraySoftware/stingray/jobs/117953343#L609) on Travis but it works on appveyor, which is also the expected behavior.

OrkoHunter commented:
Δ changes == 1.4%  =)

PS. I think there is no need of coveralls to comment after I refresh every time. We can see the stats right here.

matteobachetti commented:
@OrkoHunter I think I eliminated the automatic comments now. They are annoying, I agree :)

OrkoHunter commented:
Green ticks :tada:

OrkoHunter commented:
I created a Jupyter Notebook based on the current `Lightcurve` and changes introduced by this PR.
It can be seen on [nbviewer](http://nbviewer.jupyter.org/github/OrkoHunter/notebooks/blob/master/Stingray/Lightcurve%20tutorial.ipynb).

@dhuppenkothen @matteobachetti Can you please review this PR now, and suggest changes?

matteobachetti commented:
@OrkoHunter looks good! 
One possible change I would personally make to this is to return a `Lightcurve` object after slicing (either by index or time), not only the counts.

OrkoHunter commented:
@matteobachetti I agree with you. Now we can do something like this
```python
In []: lc_copy = lc[:]
```

Updated !

OrkoHunter commented:
@dhuppenkothen Are there any more comments for this PR?

dhuppenkothen commented:
@OrkoHunter @matteobachetti I think in principle this is ready for merging (except for the merge conflicts due to me merging another PR earlier). 

However, on a more fundamental level, there was some discussion a couple of weeks ago wether a Lightcurve class should be using astropy.table under the hood, which might simplify a lot of this? 
Open for discussion.

dhuppenkothen commented:
@OrkoHunter I made a new repository within the StingraySoftware org for tutorial and example notebooks: https://github.com/StingraySoftware/notebooks/tree/master
Feel free to submit a PR with your notebook. I think we should probably use the code in this repo https://github.com/adrn/DropOutput to strip output/row numbers before submitting, otherwise making updates to the notebook will inflate the repo a lot as we add more stuff.

OrkoHunter commented:
Rebased and fixed the merge conflicts.

matteobachetti commented on stingray/lightcurve.py:
```diff
@@ -185,3 +332,266 @@ def rebin_lightcurve(self, dt_new, method='sum'):
+        """
+        if self.dt != other.dt:
+            utils.simon("The bin widths of both the lightcurves are not "
+                        "same.")
```
"The two light curves have different bin widths"

matteobachetti commented on stingray/lightcurve.py:
```diff
@@ -185,3 +332,266 @@ def rebin_lightcurve(self, dt_new, method='sum'):
+                        "For the common time range, resulting count will be "
+                        "the average of the counts in both lightcurves. If "
+                        "you wish to sum, use `lc_sum = lc1 + lc2`.")
+
```
"The two light curves have overlapping time ranges.
In the common time range, THE resulting counts will be the average of the counts in the two light curves. (...)"

OrkoHunter commented:
@matteobachetti Previous two comments addressed.

matteobachetti commented:
@dhuppenkothen @OrkoHunter about using Astropy's Table, I think it would indeed simplify something. However, I'm finding sometimes performance issues when dealing with Astropy `Column`s instead of numpy `array`s. I would give the topic a good and lengthy discussion, and merge the current setup as it is.

dhuppenkothen commented:
@OrkoHunter @matteobachetti So accept the PR as is now and perhaps later refactor if we decide to use astropy.table?

matteobachetti commented:
Yep

On 17:15, Thu, Apr 7, 2016 Daniela Huppenkothen <notifications@github.com>
wrote:

> @OrkoHunter <https://github.com/OrkoHunter> @matteobachetti
> <https://github.com/matteobachetti> So accept the PR as is now and
> perhaps later refactor if we decide to use astropy.table?
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly or view it on GitHub
> <https://github.com/StingraySoftware/stingray/pull/74#issuecomment-206948669>
>

dhuppenkothen commented:
All right. Merging this. Thanks, @OrkoHunter, this is a fantastic addition to the code base!

OrkoHunter commented:
@evandromr @matteobachetti @dhuppenkothen Thanks for the reviews !
